### PR TITLE
Fix: repository_files subtest fails during IT (#4074, #4093)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Connected issue: #4093 

### Author

- [x] Target branch is `develop`
- [x] Source branch is `issues/<github-handle>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                    <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after rejection)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Ticket is in *Review requested* column
- [x] Requested review from peer reviewer
- [x] Assigned PR to peer reviewer

### Peer reviewer (after rejection)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] `N reviews` label on PR reflects prior peer reviews with changes requested
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Branch pushed to GitLab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `chain`</sub>
- [ ] Pushed merge commit to GitLab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from GitHub and GitLab
- [ ] Build passes on GitLab

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: test/integration_test.py [hunk 1]
```diff
@@ -750,11 +750,24 @@ def _test_repository_files(self, catalog: str):
                     response = self._get_url_json(response['Location'])
 
                 response = self._get_url(response['Location'], stream=True)
-                self._validate_file_response(response, file['format'])
+                self._validate_file_response(response, self._file_format(file))
+
+    def _file_format(self, file: JSON) -> str:
```
Comment (hannes-ucsc) [inline test/integration_test.py:755]: Let's not call this "format" to avoid confusion with the metadata property of the same name.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Let's solve #4074 in this PR and close the one currently open for #4074.

Connect this PR to both issues in Zenhub. I'll leave it to you whether the branch has two or one commit.

**Review by hannes-ucsc — COMMENTED** [review]
Build wasn't green when I tried to review this.

File: test/integration_test.py [hunk 2]
```diff
@@ -745,18 +750,37 @@ def _test_repository_files(self, catalog: str):
...
-                self._validate_fastq_response(response)
+                self._validate_file_response(response, self._file_ext(file))
+
+    def _file_ext(self, file: JSON) -> str:
+        # We believe that the file extension is a more reliable indicator than
+        # the `format` metadata field. Unlike os.path.splitext, this method
+        # preserves multipart extensions, so the extension of "foo.fastq.gz" is
+        # "fastq.gz" instead of "gz"
+        file_name = file['name']
+        _, sep, ext = file_name.lstrip('.').partition('.')
```
Comment (hannes-ucsc) [inline test/integration_test.py:761]:
```python
>>> from pathlib import PurePath
>>> PurePath('.x.y').suffix
'.y'
```

File: test/integration_test.py [hunk 3]
```diff
@@ -745,18 +750,37 @@ def _test_repository_files(self, catalog: str):
...
+        assert sep == '.', file_name
+        return ext.lower()
+
+    def _validate_file_content(self, content: ReadableFileObject, ext: str):
```
Comment (hannes-ucsc) [inline test/integration_test.py:765]:
```suggestion
    def _validate_file_content(self, content: ReadableFileObject, file_ext: str):
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Rest looks good.

File: test/integration_test.py (hunk 2 above)
Comment (noah-aviel-dove) [reply test/integration_test.py:761]: Agreed at PL that `suffixes` is the correct method to use.

**Review by hannes-ucsc — APPROVED** [review]

**Review by achave11 — APPROVED** [review]
Nicely done!

**Pull request merged**